### ISSUE TITLE
Regression: 4.1.0 broke with(csrf())

### ISSUE THREAD

**Issue by theHacker** [spring-projects/spring-security#3881]
### Summary

Since version 4.1.0 tests using `with(csrf())` are failing because the `CsrfFilter` does not get the token anymore and therefore an unexpected `AccessDeniedException` is thrown.

### Actual Behavior

Tests using `with(csrf())` are getting a status code 403.

I dug a little deeper. The problems seems to be the introduction of `TestCsrfTokenRepository`.
https://github.com/spring-projects/spring-security/blame/d3a9cc6eae56fb9826d86f021e71faa482675abe/test/src/main/java/org/springframework/security/test/web/servlet/request/SecurityMockMvcRequestPostProcessors.java#L323

`SecurityMockMvcRequestPostProcessors.postProcessRequest()` wrapps the `repository` with a `TestCsrfTokenRepository`.
When `saveToken()` is called the `TestCsrfTokenRepository` only attaches the token to the request as an attribute. But it does not delegate the call to the `HttpSessionCsrfTokenRepository` to save the token onto the session.
`CsrfFilter`'s `tokenRepository` is the `HttpSessionCsrfTokenRepository`. When the request is filtered, a lookup is done in the session, but there is no token, so `AccessDeniedException` is thrown and the test ends up with status code 403.

### Expected Behavior

`CsrfFilter` gets the token correctly, when test uses `with(csrf())`.

### Configuration

Nothing special. Spring Security enables CSRF protection by default.

### Version

- 4.0.2, 4.0.3, 4.0.4: not affected
- 4.1.0.RC1, 4.1.0.RELEASE: affected

### Sample

    @Test
    public void testMe() throws Exception {
        MyController myController = new MyController();

        MockMvc mockMvc = MockMvcBuilders
                .standaloneSetup(myController)
                .addFilters(springSecurityFilterChain)
                .build();

        mockMvc.perform(post("/foo.html")     // Spring Security default requires a CSRF token for POST
                .with(csrf()))                // Yes, we have a valid token
                .andExpect(status().is(302)); // expect redirection since we are not logged in
    }

I created a full sample project: https://github.com/theHacker/spring-security-4.1.0-csrf-bug
In `pom.xml` switch the version of Spring Security to enable/disable the bug.

**Comment by rwinch** [comment]
Thank you for such a detailed write up!

**Comment by rwinch** [comment]
@theHacker I have created a pull request with a fix for the team to review. I wanted to let you know that the fix is slightly different than your proposal.

We do not want the test support invoking the original implementation so the changes you see are quite intentional. Part of the problem is that you are using a standalone stetup vs `webAppContextSetup`. If you switch to something like the following things will work as expected:

```java
this.mockMvc = MockMvcBuilders
	.webAppContextSetup(this.context)
	.apply(springSecurity())
	.build();
```

This might be a viable option since you are already creating the ApplicationContext anyways.

The changes I introduced will fix for the standalone setup. However, you MUST use `apply(springSecurity())`. Specifically the following will NOT work:

```java
this.mockMvc = MockMvcBuilders
	.standaloneSetup(controller)
	.addFilter(springSecurityFilterChain) // this is not correct
	.build();
```

The following will work after my changes are merged

```java
this.mockMvc = MockMvcBuilders
	.standaloneSetup(controller)
	.apply(springSecurity(springSecurityFilterChain))
	.build();
```
